### ISSUE TITLE
Bad ValueType field

### ISSUE THREAD

**Issue by liqunl** [eclipse-openj9/openj9#10506]
@Leonardo2718 found a crash in his local ValueType test. The problem is caused by a bad pointer in a field of a ValueType object.

```
(kca) obj 0x00000007ffe4fdb0
Object is of type:
   Class Path/Name: {LargeObject} J9Class 0x00000000003a1000
Hidden Fields:
Super class fields: {java/lang/Object} Number of Fields: 0
Class fields:  Number of Fields: 16
>                           val1 ( +4) = 0x00000007ffe05eb8 (QValueObject;) - 20011 Public Final
>                           val2 ( +8) = 0x00000007fffff850 (QValueObject;) - 20011 Public Final     <----  bad pointer
>                           val3 (+12) = 0x00000007ffe05e98 (QValueObject;) - 20011 Public Final
>                           val4 (+16) = 0x00000007ffe05e88 (QValueObject;) - 20011 Public Final
>                           val5 (+20) = 0x00000007ffe05e78 (QValueObject;) - 20011 Public Final
>                           val6 (+24) = 0x00000007ffe05e68 (QValueObject;) - 20011 Public Final
>                           val7 (+28) = 0x00000007ffe05e58 (QValueObject;) - 20011 Public Final
>                           val8 (+32) = 0x00000007ffe05e48 (QValueObject;) - 20011 Public Final
>                           val9 (+36) = 0x00000007ffe05e38 (QValueObject;) - 20011 Public Final
>                          val10 (+40) = 0x00000007ffe05e28 (QValueObject;) - 20011 Public Final
>                          val11 (+44) = 0x00000007ffe05e18 (QValueObject;) - 20011 Public Final
>                          val12 (+48) = 0x00000007ffe05e08 (QValueObject;) - 20011 Public Final
>                          val13 (+52) = 0x00000007ffe05df8 (QValueObject;) - 20011 Public Final
>                          val14 (+56) = 0x00000007ffe05de8 (QValueObject;) - 20011 Public Final
>                          val15 (+60) = 0x00000007ffe05dd8 (QValueObject;) - 20011 Public Final
>                          val16 (+64) = 0x00000007ffe05dc8 (QValueObject;) - 20011 Public Final

```
Note that kca dump the fields uncompressed.

I have looked at places where QValueObject is created and assigned to the field, find nothing. I notice that `0x00000007fffff850` is in java heap, but it is too far away from objects in other fields, which are created at the same time.

As we can see, object of other fields are 16 bytes apart, so the expected value for `val2` is `0x00000007ffe05ea8`. And `0x00000007ffe05ea8` is a valid ValueObject
```
(kca) what 0x00000007ffe05ea8
Obj - {ValueObject}
```

```
1STHEAPTYPE    Object Memory
NULL           id                 start              end                size               space/region
1STHEAPSPACE   0x00007F619C0934B0         --                 --                 --         Generational 
1STHEAPREGION  0x00007F619C094120 0x00000006140E0000 0x00000006146E0000 0x0000000000600000 Generational/Tenured Region 
1STHEAPREGION  0x00007F619C093B50 0x00000007FFE00000 0x00000007FFFA0000 0x00000000001A0000 Generational/Nursery Region 
1STHEAPREGION  0x00007F619C093580 0x00000007FFFA0000 0x0000000800000000 0x0000000000060000 Generational/Nursery Region 
```
From javacore, we have two nursery spaces, current fields are in the first nursery, and the bad pointer is in the second nursery.
The heap allocation range in VMThread is the following, which is in the first nursery
```
        0x60: U8* heapAlloc = !j9x 0x00000007FFE5D3A0
        0x68: U8* heapTop = !j9x 0x00000007FFE5FA08
```
So it seems like the objects have been moved, but something went wrong in updating `val2`.

Don't know if it's some traces left by gc moving the objects, the bad pointer `0x00000007fffff850` points to the middle of a ValueObject `0x00000007ffe05ea8` (0x00000007ffe05ea8 + 4), which is the expected one for `val2`.
```
(kca) what (0x00000007fffff850)/l
0x7fffff850: 0x00000007ffe05eac Obj - {ValueObject} +4
```

The crash hasn't been reproduced yet, so the core file is the only thing we got.
@dmitripivkine @amicic I wonder if you or someone in GC team lend us a hand? The JIT has run out of places to look and it seems like something related to GC.

**Comment by dmitripivkine** [comment]
@liqunl  Would you please let me know where I can find system core?

**Comment by dmitripivkine** [comment]
I appreciate your question and want to encourage you to ask for help from GC Team in cases like this.

There is obvious that "bad pointer" refer to reserved (wrong) half of Nursery. It is not actually "bad" but Forwarded Pointer (with added Forwarded flag 0x4):
```
0x7FFFFF850 :  ffe05eac 00000007 00000000 00000000 [ .^.............. ] <-- FP to 0x7ffe05ea8
0x7FFFFF860 :  00399600 00000000 00000000 00000000 [ ..9............. ]

> !j9object 0x7ffe05ea8
!J9Object 0x00000007FFE05EA8 {
	struct J9Class* clazz = !j9class 0x399600 // ValueObject
	Object flags = 0x00000010;
	Ljava/lang/Object; val = !fj9object 0xc2854341 (offset = 0) (ValueObject)
}
```
An array itself `!j9object 0x7FFE4FDB0` and all other elements located it the correct size of Nursery. It means that at the time of Scavenge an array has been copied, scanned and its elements have been copied too. It means that problematic address was not there at that time (pointed to another object or NULL) and was installed later on after Scavenge. And this object address has not been fixed up by GC so it is stall pointer. So the are two possibilities to fix the problem:
- prevent GC to occur mid operation by keeping VM Access
- make temporary location of object address be visible for GC to fix it up

**Comment by liqunl** [comment]
Thanks for helping out. With an offline discussion with @dmitripivkine, we concluded that `LargetObject` is good after construction, and its fields are good right after Scavenge. After Scavenge, there should be a method that writes to `val2` with value from a temporary location (register or stack slot) where an address of this object was stored which has not been fixed up. And likely that temporary location is not visible to GC (not in gc map).

I initially thought the field is bad when `LargeObject` is constructed because the test doesn't seem to be modifying the fields. I'll read the test again to see if I can find a method that writes to the field.

**Comment by liqunl** [comment]
Should be fixed by #10976

**Comment by dmitripivkine** [comment]
I think this issue can be closed
